components/ca-treeview.js

```diff
@@ -0,0 +1,263 @@
+            }
+        }
+
+        /**
+         * Called every time an attribute is changed.
+         * @param {string} attrName, the name of the changed attribute.
+         * @param {string} oldVal, previous value associated with attribute.
+         * @param {string} newVal, current value associated with attribute.
+         * @returns {undefined} if data-src or data-root change, then trigger data fetch.
+         */
+        attributeChangedCallback(attrName, oldVal, newVal) {
+            switch (attrName) {
+                case 'data-src': {
+                    if (newVal !== '') {
+                        this.loadData();
+                    }
+                } break;
+
+                case 'data-root': {
+                    this.root = newVal;
+                } break;
+
+                default: break;
+            }
+        }
+
+        /**
+         * Gets the root attribute
+         * @returns {string} data-root attribute, default is empty.
+         */
+        get root() {
+            return this.getAttribute('data-root') || '';
+        }
+
+        /**
+         * Gets the root attribute
+         * @param {string} value, data-root attribute, default is empty.
+         */
+        set root(value = '') {
+            this.setAttribute('data-root', value);
+        }
+
+        /**
+         * Gets the data-src
+         * @returns {string} data-src attribute.
+         */
+        get dataSrc() {
+            return this.getAttribute('data-src');
+        }
+
+        /**
+         * Sets the data-src
+         * @param {string} value, sets the data-src attribute.
+         */
+        set dataSrc(value) {
+            this.setAttribute('data-src', value);
+        }
+
+        /**
+         * Get the data
+         * @returns {array} returns the data.
+         */
+        get data() {
+            return this._data || [];
+        }
+
+        /**
+         * Sets the data
+         * @param {array} value, array to set as the data, defaults to empty. Invokes render.
+         */
+        set data(value = []) {
+            this._data = value;
+            this.render();
+        }
+
+        /**
+         * Gets selected item
+         * @returns {HTMLElement} selected Item
+         */
+        get selected() {
+            return this._selected;
+        }
+
+        /**
+         * Sets the selected item.
+         * @param {HTMLElement} el current element.
+         * @returns {int} max-children.
+         */
+        set selected(el) {
+            this._selected = el;
+        }
+
+        /**
+         * Gets the max children allowed.
+         * @returns {int} max-children.
+         */
+        get maxChildren() {
+            return parseInt(this.getAttribute('data-max-children') || '0', 10);
+        }
+
+        /**
+         * Sets the max children allowed.
+         * @param {int} value number of allow childre, defaults to 0.
+         */
+        set maxChildren(value = 0) {
```

Comment:
Also, document.querySelector('body').setAttribute('baaa-baaa-moewww', 5) -> would produce <body baaa-baaa-moewww="5>..</body>

And fetching it document.querySelector('body').getAttribute('baaa-baaa-moewww') -> "5"